ext/openssl/openssl_pwhash.c

```diff
@@ -0,0 +1,396 @@
+   +----------------------------------------------------------------------+
+   | Authors: Remi Collet <remi@php.net>                                  |
+   +----------------------------------------------------------------------+
+*/
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include "php.h"
+#include "ext/standard/php_password.h"
+#include "php_openssl.h"
+
+#if defined(HAVE_OPENSSL_ARGON2)
+#include "Zend/zend_attributes.h"
+#include "openssl_pwhash_arginfo.h"
+#include <ext/standard/base64.h>
+#include "ext/random/php_random_csprng.h"
+#include <openssl/params.h>
+#include <openssl/core_names.h>
+#include <openssl/kdf.h>
+#include <openssl/thread.h>
+#include <openssl/rand.h>
+
+#define PHP_OPENSSL_MEMLIMIT_MIN  8u
+#define PHP_OPENSSL_MEMLIMIT_MAX  UINT32_MAX
+#define PHP_OPENSSL_ITERLIMIT_MIN 1u
+#define PHP_OPENSSL_ITERLIMIT_MAX UINT32_MAX
+#define PHP_OPENSSL_THREADS_MIN   1u
+#define PHP_OPENSSL_THREADS_MAX   UINT32_MAX
+
+#define PHP_OPENSSL_ARGON_VERSION 0x13
+
+#define PHP_OPENSSL_SALT_SIZE     16
+#define PHP_OPENSSL_HASH_SIZE     32
+#define PHP_OPENSSL_DIGEST_SIZE  128
+
+static inline zend_result get_options(zend_array *options, uint32_t *memlimit, uint32_t *iterlimit, uint32_t *threads)
+{
+	zval *opt;
+
+	*iterlimit = PHP_OPENSSL_PWHASH_ITERLIMIT;
+	*memlimit  = PHP_OPENSSL_PWHASH_MEMLIMIT;
+	*threads   = PHP_OPENSSL_PWHASH_THREADS;
+
+	if (!options) {
+		return SUCCESS;
+	}
+	if ((opt = zend_hash_str_find(options, "memory_cost", strlen("memory_cost")))) {
+		zend_long smemlimit = zval_get_long(opt);
+
+		if ((smemlimit < 0) || (smemlimit < PHP_OPENSSL_MEMLIMIT_MIN) || (smemlimit > (PHP_OPENSSL_MEMLIMIT_MAX))) {
+			zend_value_error("Memory cost is outside of allowed memory range");
+			return FAILURE;
+		}
+		*memlimit = smemlimit;
+	}
+	if ((opt = zend_hash_str_find(options, "time_cost", strlen("time_cost")))) {
+		zend_long siterlimit = zval_get_long(opt);
+		if ((siterlimit < PHP_OPENSSL_ITERLIMIT_MIN) || (siterlimit > PHP_OPENSSL_ITERLIMIT_MAX)) {
+			zend_value_error("Time cost is outside of allowed time range");
+			return FAILURE;
+		}
+		*iterlimit = siterlimit;
+	}
+	if ((opt = zend_hash_str_find(options, "threads", strlen("threads"))) && (zval_get_long(opt) != 1)) {
+		zend_long sthreads = zval_get_long(opt);
+		if ((sthreads < PHP_OPENSSL_THREADS_MIN) || (sthreads > PHP_OPENSSL_THREADS_MAX)) {
+			zend_value_error("Invalid number of threads");
+			return FAILURE;
+		}
+		*threads = sthreads;
+	}
+	return SUCCESS;
+}
+
+static bool php_openssl_argon2_compute_hash(
+	const char *algo,
+	uint32_t version, uint32_t memlimit, uint32_t iterlimit, uint32_t threads,
+	const char *pass, size_t pass_len,
+	const unsigned char *salt, size_t salt_len,
+	unsigned char *hash, size_t hash_len)
+{
+	OSSL_PARAM params[7], *p = params;
+	EVP_KDF *kdf = NULL;
+	EVP_KDF_CTX *kctx = NULL;
+	bool ret = false;
+
+	if (threads > 1) {
+		if (OSSL_set_max_threads(NULL, threads) != 1) {
+			goto fail;
+		}
+	}
```

Comment:
I think it would be better going forward to not relay that max threads is always 0 in case this gets used from elsewhere (I guess even 3rd party extension or lib could maybe set it). So it might be better to first get the original max threads (`OSSL_get_max_threads`) and then re-set it after the hashing.

Reply:
done